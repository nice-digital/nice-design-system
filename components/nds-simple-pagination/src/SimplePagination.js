import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import "./../scss/simple-pagination.scss";

const Link = ({ text, destination, elementType: LinkTag = "a" }) => {
	let linkProps = {
		className: "simple-pagination__link"
	};
	if (LinkTag === "a") {
		linkProps.href = destination;
	} else {
		linkProps.to = destination;
	}
	return (
		<span className="simple-pagination__link-wrapper">
			<LinkTag {...linkProps}>{text}</LinkTag>{" "}
		</span>
	);
};

export const SimplePagination = props => {
	const {
		currentPage,
		totalPages,
		nextPageLink,
		previousPageLink,
		className,
		...rest
	} = props;

	const nextLinkProps = {
		text: "Next page",
		destination: nextPageLink && nextPageLink.destination,
		elementType: nextPageLink && nextPageLink.elementType
	};

	const previousLinkProps = {
		text: "Previous page",
		destination: previousPageLink && previousPageLink.destination,
		elementType: previousPageLink && previousPageLink.elementType
	};

	return (
		<div className={classnames("simple-pagination", className)} {...rest}>
			{currentPage && (
				<p>
					Page <b>{currentPage}</b>
					{totalPages && (
						<>
							{" "}
							of <b>{totalPages}</b>
						</>
					)}
				</p>
			)}
			{(nextPageLink || previousPageLink) && (
				<nav aria-label="Pagination">
					{nextPageLink && <Link {...nextLinkProps} />}
					{previousPageLink && <Link {...previousLinkProps} />}
				</nav>
			)}
		</div>
	);
};

const LinkPropTypes = PropTypes.shape({
	destination: PropTypes.node,
	element: PropTypes.elementType
});

SimplePagination.propTypes = {
	currentPage: PropTypes.number.isRequired,
	nextPageLink: LinkPropTypes,
	previousPageLink: LinkPropTypes,
	totalPages: PropTypes.number
};

SimplePagination.defaultProps = {
	currentPage: 1,
	totalPages: null,
	nextPageLink: null,
	previousPageLink: null
};
