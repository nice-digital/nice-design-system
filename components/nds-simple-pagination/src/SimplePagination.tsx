import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import "./../scss/simple-pagination.scss";

interface PageLinkProps {
	destination?: string;
	elementType?: React.ElementType;
	method?: string;
	text?: string;
}

export interface SimplePaginationProps {
	[prop: string]: unknown;
	currentPage?: number;
	totalPages?: number;
	nextPageLink?: PageLinkProps;
	previousPageLink?: PageLinkProps;
	className?: string;
}

const Link: React.FC<PageLinkProps> = ({
	text,
	destination,
	elementType: ElementType = "a",
	method
}: PageLinkProps) => {
	let linkProps = {
		className: "simple-pagination__link",
		[method || (ElementType === "a" && "href") || "to"]: destination
	};

	return (
		<span className="simple-pagination__link-wrapper">
			<ElementType {...linkProps}>{text}</ElementType>{" "}
		</span>
	);
};

export const SimplePagination: React.FC<SimplePaginationProps> = (
	props: SimplePaginationProps
) => {
	const {
		currentPage,
		totalPages,
		nextPageLink,
		previousPageLink,
		className,
		...rest
	} = props;

	const nextLinkProps = {
		text: "Next page",
		destination: nextPageLink && nextPageLink.destination,
		elementType: nextPageLink && nextPageLink.elementType,
		method: nextPageLink && nextPageLink.method
	};

	const previousLinkProps = {
		text: "Previous page",
		destination: previousPageLink && previousPageLink.destination,
		elementType: previousPageLink && previousPageLink.elementType,
		method: previousPageLink && previousPageLink.method
	};

	return (
		<div className={classnames("simple-pagination", className)} {...rest}>
			{currentPage && (
				<p>
					Page <b>{currentPage}</b>
					{totalPages && (
						<>
							{" "}
							of <b>{totalPages}</b>
						</>
					)}
				</p>
			)}
			{(nextPageLink || previousPageLink) && (
				<nav aria-label="Pagination">
					{nextPageLink && <Link {...nextLinkProps} />}
					{previousPageLink && <Link {...previousLinkProps} />}
				</nav>
			)}
		</div>
	);
};
