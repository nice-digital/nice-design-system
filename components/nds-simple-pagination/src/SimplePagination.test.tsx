import React from "react";
import { render } from "@testing-library/react";

import { SimplePagination } from "./SimplePagination";

const Link = () => <a href="#">I am a link</a>;

const props = {
	totalPages: 3,
	currentPage: 1,
	nextPageLink: {
		destination: "/next",
		elementType: Link
	},
	previousPageLink: {
		destination: "#previous"
	}
};

describe("Simple Pagination", () => {
	it("should match snapshot with some default attributes", () => {
		const wrapper = render(<SimplePagination {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	it("should default to an anchor if no elementType is provided", () => {
		const { container } = render(<SimplePagination {...props} />);
		const previousLink = container.querySelector("a[href='#previous']");
		expect(previousLink).toBeInTheDocument();
	});

	it("should use a custom navigation attribute if supplied", () => {
		const localProps = Object.assign({}, props, {
			previousPageLink: {
				destination: "#previous",
				method: "pigeon"
			}
		});
		const { container } = render(<SimplePagination {...localProps} />);
		const previousLink = container.querySelector("a[pigeon='#previous']");
		expect(previousLink).toBeInTheDocument();
	});

	it("should pass extra props to the container element", () => {
		const localProps = Object.assign({}, props, {
			"data-tracker": "my-tracker"
		});
		const { container } = render(<SimplePagination {...localProps} />);
		const pagination = container.querySelector(".simple-pagination");
		expect(pagination?.getAttribute("data-tracker")).toBe("my-tracker");
	});

	it("should default to 'page 1' only if no props are provided", () => {
		const wrapper = render(<SimplePagination />);
		expect(wrapper).toMatchSnapshot();
	});

	it("should merge any additional classes on to the existing wrapper class", () => {
		const { container } = render(
			<SimplePagination className="testing hello" />
		);
		const pagination = container.querySelector(".simple-pagination");
		expect(pagination).toHaveClass("simple-pagination testing hello");
	});
});
