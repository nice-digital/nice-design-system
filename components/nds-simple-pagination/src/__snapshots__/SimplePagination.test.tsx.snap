// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple Pagination should default to 'page 1' only if no props are provided 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="simple-pagination"
        data-component="simple-pagination"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="simple-pagination"
      data-component="simple-pagination"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Simple Pagination should match snapshot with some default attributes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="simple-pagination"
        data-component="simple-pagination"
      >
        <p>
          Page 
          <b>
            1
          </b>
           
          of 
          <b>
            3
          </b>
        </p>
        <nav
          aria-label="Pagination"
        >
          <span
            class="simple-pagination__link-wrapper"
          >
            <a
              href="#"
            >
              I am a link
            </a>
             
          </span>
          <span
            class="simple-pagination__link-wrapper"
          >
            <a
              class="simple-pagination__link"
              href="#previous"
            >
              Previous page
            </a>
             
          </span>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="simple-pagination"
      data-component="simple-pagination"
    >
      <p>
        Page 
        <b>
          1
        </b>
         
        of 
        <b>
          3
        </b>
      </p>
      <nav
        aria-label="Pagination"
      >
        <span
          class="simple-pagination__link-wrapper"
        >
          <a
            href="#"
          >
            I am a link
          </a>
           
        </span>
        <span
          class="simple-pagination__link-wrapper"
        >
          <a
            class="simple-pagination__link"
            href="#previous"
          >
            Previous page
          </a>
           
        </span>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
